src/mpid/ch4/src/ch4_wait.h

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef CH4_WAIT_H_INCLUDED
+#define CH4_WAIT_H_INCLUDED
+
+#include "ch4_impl.h"
+
+MPL_STATIC_INLINE_PREFIX void MPIDI_set_progress_vci(MPIR_Request * req,
+                                                     MPID_Progress_state * state)
+{
+    state->flag = MPIDI_PROGRESS_ALL;   /* TODO: check request is_local/anysource */
+    state->progress_count = MPL_atomic_load_int(&MPIDI_global.progress_count);
```

Comment:
The current semantics of `MPID_Progress_wait` dictates that the function will return when *any* request completes processing. That's why `progress_count` is used. For per-VCI progress, however, we don't need it since we want to wait for the completion of the particular request. We could be conservative and use the current semantics but it hurts performance drastically since it is a strong source of contention.

To overcome this, I had created a new function `MPID_Progress_wait_req` which will return only when the specified request returns. That way we can get rid of the global `progress_count` atomic in the critical path for per-VCI progress.

Reply:
I understand if we let go the atomic counter and only poll a single request we are going to avoid a lot of contention. But how is it any different with per-vci progress and without?

Reply:
A per-VCI progress counter would also get rid of the contention but at the expense of fetch-and-add's overhead on the critical path. In that case, the semantics of per-VCI and global progress would be the same. I think the question boils down to: do we need the same semantics of `MPID_Progress_wait` (return when any request has been completed) for per-VCI progress? I can't think of a situation where we need this semantic.

Reply:
What about the `MPI_Waitall`? We call `MPID_Progress_wait_req` in sequence?

I don't think I disagree with change `MPID_Progress_wait` interface and supply it with an explicit request object to wait on. However, it seems the logic is the same for global progress or per-vci progress, therefore it is orthogonal?

Reply:
Similar to `MPID_Progress_wait`, `MPID_Progress_wait_req` will be called only when the request is not complete. In `MPI_Waitall`, we do iterate over every request but we call the progress engine only when the request that we are iterating over is not complete. So, if all requests in `MPI_Waitall` point to the same VCI, the first call to `MPID_Progress_wait` or `MPID_Progress_wait_req` would most likely complete all the requests, and we would not call it for every request.

Sure, introducing `MPID_Progress_wait_req` or moving the `progress_count` to a per-VCI basis could be done in another PR. Just wanted to note that we need to work around `MPIDI_global.progress_count`. Whatever works best with your plan of introducing changes :)

Reply:
Sounds good!